core/libp2p/connection/yamux/yamux_stream.cpp

```diff
@@ -0,0 +1,303 @@
+                  self->is_reading_ = false;
+                  if (!res) {
+                    return cb(res.error());
+                  }
+                  self->read_buffer_.consume(bytes);
+                  cb(bytes);
+                });
+          }
+          return true;
+        });
+  }
+
+  void YamuxStream::readSome(gsl::span<uint8_t> out, size_t bytes,
+                             ReadCallbackFunc cb) {
+    if (bytes == 0 || out.empty() || static_cast<size_t>(out.size()) < bytes) {
+      return cb(Error::INVALID_ARGUMENT);
+    }
+    if (!is_readable_) {
+      return cb(Error::NOT_READABLE);
+    }
+    if (is_reading_) {
+      return cb(Error::IS_READING);
+    }
+
+    is_reading_ = true;
+
+    // if there is already enough data in our buffer, read it immediately
+    if (read_buffer_.size() != 0) {
+      auto to_read = std::min(read_buffer_.size(), bytes);
+      if (boost::asio::buffer_copy(boost::asio::buffer(out.data(), to_read),
+                                   read_buffer_.data(), to_read)
+          != to_read) {
+        return cb(Error::INTERNAL_ERROR);
+      }
+      return yamuxed_connection_->streamAckBytes(
+          stream_id_, to_read,
+          [self{shared_from_this()}, cb = std::move(cb), to_read](auto &&res) {
+            self->is_reading_ = false;
+            if (!res) {
+              return cb(res.error());
+            }
+            self->read_buffer_.consume(to_read);
+            cb(to_read);
+          });
+    }
+
+    // else, set a callback, which is called each time a new data arrives
+    yamuxed_connection_->streamAddDataNotifyee(
+        stream_id_,
+        [self{shared_from_this()}, cb = std::move(cb), out, bytes]() mutable {
+          if (self->read_buffer_.size() == 0) {
+            // not yet
+            return false;
+          }
+
+          auto to_read = std::min(self->read_buffer_.size(), bytes);
+          if (boost::asio::buffer_copy(boost::asio::buffer(out.data(), to_read),
+                                       self->read_buffer_.data(), to_read)
+              != bytes) {
+            cb(Error::INTERNAL_ERROR);
+          } else {
+            self->yamuxed_connection_->streamAckBytes(
+                self->stream_id_, to_read,
+                [self, cb = std::move(cb), to_read](auto &&res) {
+                  self->is_reading_ = false;
+                  if (!res) {
+                    return cb(res.error());
+                  }
+                  self->read_buffer_.consume(to_read);
+                  cb(to_read);
+                });
+          }
+          return true;
+        });
+  }
+
+  void YamuxStream::write(gsl::span<const uint8_t> in, size_t bytes,
+                          WriteCallbackFunc cb) {
+    return write(in, bytes, std::move(cb), false);
+  }
+
+  void YamuxStream::writeSome(gsl::span<const uint8_t> in, size_t bytes,
+                              WriteCallbackFunc cb) {
+    return write(in, bytes, std::move(cb), true);
```

Comment:
writeSome is actually different from write, in the same way as read/readSome.

e.g. callback passed into write should be executed only if the other party acked `bytes` bytes.
callback passed into writeSome should be executed if the other party acked any number of bytes > 0 up to `bytes`

Reply:
I know. That's why I am passing`bool` into the internal `write` method - to see, if it was to write some or all bytes